backends/ubpf/examples/simple-firewall.p4

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2019 Orange
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ubpf_model.p4"
+#include <core.p4>
+
+#define SYNSENT 1
+#define SYNACKED 2
+#define ESTABLISHED 3
+
+typedef bit<48>  EthernetAddress;
+typedef bit<9> egressSpec_t;
+
+header Ethernet_t {
+    EthernetAddress dstAddr;
+    EthernetAddress srcAddr;
+    bit<16>         etherType;
+}
+
+header Ipv4_t {
+    bit<4>  version;
+    bit<4>  ihl;
+    bit<8>  diffserv;
+    bit<16> totalLen;
+    bit<16> identification;
+    bit<3>  flags;
+    bit<13> fragOffset;
+    bit<8>  ttl;
+    bit<8>  protocol;
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+header Tcp_t {
+    bit<16> srcPort;
+    bit<16> dstPort;
+    bit<32> seqNo;
+    bit<32> ackNo;
+    bit<4>  dataOffset;
+    bit<3>  res;
+    bit<3>  ecn;
+    bit<1> urgent;
+    bit<1> ack;
+    bit<1> psh;
+    bit<1> rst;
+    bit<1> syn;
+    bit<1> fin;
+    bit<16> window;
+    bit<16> checksum;
+    bit<16> urgentPtr;
+}
+
+struct Headers_t {
+    Ethernet_t       ethernet;
+    Ipv4_t           ipv4;
+    Tcp_t            tcp;
+}
+
+struct ConnectionInfo_t {
+    bit<32> s;
+    bit<32> srv_addr;
+}
+
+struct metadata {
+    ConnectionInfo_t connInfo;
+    bit<32> conn_id;
+}
+
+parser prs(packet_in packet, out Headers_t hdr, inout metadata meta) {
+    state start {
+        transition parse_ethernet;
+    }
+    state parse_ethernet {
+        packet.extract(hdr.ethernet);
+        transition select(hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        packet.extract(hdr.ipv4);
+        transition parse_tcp;
+    }
+    state parse_tcp {
+        packet.extract(hdr.tcp);
+        transition accept;
+    }
+}
+
+control pipe(inout Headers_t hdr, inout metadata meta) {
+
+    Register<bit<32>, bit<32>>(65536) conn_state;
+    Register<bit<32>, bit<32>>(65536) conn_srv_addr;
+
+    action update_conn_state(bit<32> s) {
+        conn_state.write(meta.conn_id, s);
+    }
+
+    action update_conn_info(bit<32> s, bit<32> addr) {
+        conn_state.write(meta.conn_id, s);
+        conn_srv_addr.write(meta.conn_id, addr);
+    }
+
+    action _drop() {
+        mark_to_drop();
+    }
+
+    apply {
+
+        if (hdr.tcp.isValid()) {
+            @atomic {
```

Comment:
you have implemented support for atomic?
(haven't got to that part of the code yet)

Reply:
Removed.